### PR TITLE
libct: speedup process.Env handling

### PR DESCRIPTION
**Pull request by kolyshkin**
This is a rework/carry of #1983.

----
    The current implementation sets all the environment variables passed in
    Process.Env in the current process, one by one, then uses os.Environ to
    read those back.
    
    As pointed out in [1], this is slow, as runc calls os.Setenv for every
    variable, and there may be a few thousands of those. Looking into how
    os.Setenv is implemented, it is indeed slow, especially when cgo is
    enabled.
    
    Looking into why it was implemented the way it is, I found commit
    9744d72c and traced it to [2], which discusses the actual reasons.
    It boils down to these two:
    
     - HOME is not passed into container as it is set in setupUser by
       os.Setenv and has no effect on config.Env;
     - there is a need to deduplication of environment variables.
    
    Yet it was decided in [2] to not go ahead with this patch, but
    later [3] was opened with the carry of this patch, and merged.
    
    Now, from what I see:
    
    1. Passing environment to exec is way faster than using os.Setenv and
       os.Environ (tests show ~20x speed improvement in a simple Go test,
       and ~3x improvement in real-world test, see below).
    2. Setting environment variables in the runc context may result is some
       ugly side effects (think GODEBUG, LD_PRELOAD, or _LIBCONTAINER_*).
    3. Nothing in runtime spec says that the environment needs to be
       deduplicated, or the order of preference (whether the first or the
       last value of a variable with the same name is to be used). We should
       stick to what we have in order to maintain backward compatibility.
    
    So, this patch:
     - switches to passing env directly to exec;
     - adds deduplication mechanism to retain backward compatibility;
     - takes care to set PATH from process.Env in the current process
       (so that supplied PATH is used to find the binary to execute),
       also to retain backward compatibility;
     - adds HOME to process.Env if not set.
    
    The benchmark added by the previous commit shows ~3x improvement:
    
                            │   before    │                after                 │
                            │   sec/op    │    sec/op     vs base                │
            ExecInBigEnv-20   61.53m ± 1%   21.87m ± 16%  -64.46% (p=0.000 n=10)

[1]: https://github.com/opencontainers/runc/pull/1983
[2]: https://github.com/docker-archive/libcontainer/pull/418
[3]: https://github.com/docker-archive/libcontainer/pull/432

----
    
Remaining questions (and my answers to those):

Q: Are there any potential regressions (for example, from not setting values from `process.Env` to the current process?
A: Pprobably not; if yes, someone is exploiting some undocumented behavior.

Q: Should deduplication show warnings (maybe promoted to errors later)?
A: For best backward compatibility, let's not do that. Can always be added later (maybe with some addition to runtime-spec).

Q: Whether a default for `PATH` (e.g `"/bin:/usr/bin"` should be added, when `PATH` is not set.
A: This needs to be done in runtime-spec first (document the default for PATH, then add it to runtimes).

### CODE REVIEW

File: libcontainer/integration/execin_test.go [hunk 1]
```diff
@@ -377,6 +378,73 @@ func TestExecInEnvironment(t *testing.T) {
...
+	for i := 0; i < count; i++ {
+		key := strings.ToUpper(randStr(10))
+		value := randStr(20)
+		envs[i] = key + "=" + value
+	}
+
+	return envs
+}
+
+func BenchmarkExecInBigEnv(b *testing.B) {
```
Comment (lifubang) [inline libcontainer/integration/execin_test.go:401]: Would you remove these benchmark test code later?

File: libcontainer/env.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+// prepareEnv checks supplied environment variables for validity, removes
+// duplicates (leaving the last value only), and sets PATH from env, if found.
+// Returns the deduplicated environment, and a flag telling if HOME is found.
+func prepareEnv(env []string) ([]string, bool, error) {
+	if env == nil {
+		return nil, false, nil
+	}
+	// Deduplication code based on dedupEnv from Go 1.22 os/exec.
```
Comment (lifubang) [inline libcontainer/env.go:18]: Maybe we should not remove the duplicated env in runc, just let Go or kernel to handle it?

File: libcontainer/init_linux.go [hunk 3]
```diff
@@ -196,10 +195,6 @@ func startInitialization() (retErr error) {
 		dmzExe = os.NewFile(uintptr(dmzFd), "runc-dmz")
 	}
 
-	// clear the current process's environment to clean any libcontainer
-	// specific env vars.
-	os.Clearenv()
```
Comment (lifubang) [inline libcontainer/init_linux.go:201]: How about remaining this? Is this affect the performance?

File: libcontainer/integration/execin_test.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/integration/execin_test.go:401]: Why should I? What's wrong in leaving it?

File: libcontainer/init_linux.go (hunk 3 above)
Comment (kolyshkin) [reply libcontainer/init_linux.go:201]: Unfortunately it does affect performance. What it does internally is calls Unsetenv for every variable (and Unsetenv calls glibc's `unsetenv(3)`). You can add this back and use the benchmark provided to see what difference does it make (I haven't checked it but I checked Clearenv in other benchmarks and it is slow).

Also, while I understand the security concert, with this PR clearenv is definitely not needed because now the container process does not use any environment from the parent process.

Comment (cyphar) [reply libcontainer/init_linux.go:201]: Damn you're right, Go has their own copy of the environment and they implement it that way (`clearenv(3)` is very efficient by comparison). It seems the reason they clear the environment that way is because they need to detect changes to `GODEBUG` at runtime -- but this could be done more efficiently (by clearing the environment properly and then just checking whether the old environment had a `GODEBUG` value and then calling the unset handler for just that value).

I don't think there is a practical issue with dropping it, but I would prefer to fix the Go stdlib issue rather than dropping this. I don't want some other change to result in us calling out to a binary and we accidentally leak environment variables from the host.

<hr>

Funnily enough, this code was originally added in 44636dd8632756a70400793d84aeab3673a8d057 AFAICS, where the way they did re-exec at the time meant they moved from using `command.Env` to `Clearenv`+`Setenv`. :smile_cat:

File: libcontainer/integration/execin_test.go (hunk 1 above)
Comment (lifubang) [reply libcontainer/integration/execin_test.go:401]: https://github.com/opencontainers/runc/blob/cda4c42056dac11727867726aec08c3d88f3d1a0/libcontainer/integration/execin_test.go#L442

File: libcontainer/standard_init_linux.go [hunk 4]
```diff
@@ -277,7 +278,7 @@ func (l *linuxStandardInit) Init() error {
 
 	if l.dmzExe != nil {
 		l.config.Args[0] = name
-		return system.Fexecve(l.dmzExe.Fd(), l.config.Args, os.Environ())
+		return system.Fexecve(l.dmzExe.Fd(), l.config.Args, l.config.Env)
```
Comment (lifubang) [inline libcontainer/standard_init_linux.go:281]: Before we merge #4323 , maybe we should include `l.config.Env` when running `StartContainer` hook.

File: libcontainer/init_linux.go (hunk 3 above)
Comment (lifubang) [reply libcontainer/init_linux.go:201]: Yes, I also prefer to keep this, because there are usually only no more than 10 envs here.

File: libcontainer/integration/execin_test.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/integration/execin_test.go:401]: Ah! My bad. I wanted some way to check the output and did not finish it.

This is now updated.

File: libcontainer/standard_init_linux.go (hunk 4 above)
Comment (kolyshkin) [reply libcontainer/standard_init_linux.go:281]: This is definitely something to think about. Maybe it makes sense to do it selectively (for those hooks that are run inside the container -- AFAIR not all of them are)

**Comment by thaJeztah** [comment]
> Nothing in runtime spec says that the environment needs to be deduplicated, or the order of preference (whether the first or the last value of a variable with the same name is to be used). In C (Linux/glibc), the first value is used. In Go, it's the last one.

Interesting; I didn't know about that discrepancy; sounds like something that wouldn't hurt to define in the OCI spec; given that all original implementations were in Go, and ISTR Docker also had its own code to remove duplicates, I'm inclined to describe that as the expected behavior (possibly recommend producers of the OCI config to handle duplicates themselves to prevent any ambiguity).

**Comment by kolyshkin** [comment]
> Nothing in runtime spec says that the environment needs to be deduplicated, or the order of preference (whether the first or the last value of a variable with the same name is to be used). In C (Linux/glibc), the first value is used. In Go, it's the last one.

I was not entirely correct here. Let me rephrase this:

1. When the environment is set using glibc's `setenv(3)` or `putenv(3)`, or Go's `os.Putenv`, the value from the latter call for a given key takes precedence. This is obvious.
2. If the existing environment is not deduplicated (such as when you supply it directly to execve), both glibc's `getenv(3)` and Go's `os.Getenv` return the value of the first element.
3. When using a higher level primitives (from `os/exec`) in Go, the environment passed is deduplicated in a way that the last value of a key is used. In glibc, there is no such deduplication (as far as I can see), and thus the first value of the key will be used. This is the difference between Go and C.

----

On an unrelated note, I also took a look at crun and it seems it is following runc logic, calling `clearenv(3)` followed by `putenv(3)` for every item in `Process.Env`, when using `environ` for execve -- which result in the latter value of the same env var being used. There is probably no room for a similar optimization as `putenv` in glibc is not thread safe and there are no mutexes to be held etc.

**Comment by kolyshkin** [comment]
PTAL @opencontainers/runc-maintainers

**Comment by lifubang** [comment]
> PTAL @opencontainers/runc-maintainers

LGTM, it’s a big step.
But as I mentioned in https://github.com/opencontainers/runc/pull/4325#discussion_r1656699472, I suggest to add these envs to hook’s process execution before we have a conclusion for #4323.

File: libcontainer/env.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+	// Construct the output in reverse order, to preserve the
+	// last occurrence of each key.
+	out := make([]string, 0, len(env))
+	saw := make(map[string]bool, len(env))
+	for n := len(env); n > 0; n-- {
+		kv := env[n-1]
```
Comment (rata) [inline libcontainer/env.go:28]: nit: It seems simpler to use n = len(env) -1, to keep the structure of loops we used for a lot of years already (but you need to enter on the n=o case too)

**Review by rata — COMMENTED** [review]
@kolyshkin left some comments, but seems great it makes such a huge difference! This is marked for 1.3, though, but I guess you asking for reviews now means you want this for 1.2?

I'm fine with this in 1.2.

File: libcontainer/env.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		if saw[key] { // Duplicate.
+			continue
+		}
+		saw[key] = true
+		if strings.IndexByte(kv, 0) >= 0 {
+			return nil, false, fmt.Errorf("invalid environment variable %q: contains nul byte (\\x00)", key)
+		}
+		if key == "PATH" {
+			// Needs to be set as it is used for binary lookup.
+			if err := os.Setenv("PATH", kv[5:]); err != nil {
```
[the remark this answers is not in the document]
Comment (rata) [reply libcontainer/env.go:46]: I'm unsure about doiong this here vs in the system.Exec() func. The latter seems more tidy, IMHO. I guess it is only needed there, right? Or maybe some hooks that exec before the container need it too?

File: libcontainer/env.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+	// Construct the output in reverse order, to preserve the
+	// last occurrence of each key.
+	out := make([]string, 0, len(env))
+	saw := make(map[string]bool, len(env))
+	for n := len(env); n > 0; n-- {
+		kv := env[n-1]
+		i := strings.IndexByte(kv, '=')
+		if i == -1 {
+			return nil, false, errors.New("invalid environment variable: missing '='")
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply libcontainer/env.go:31]: We decided to not do it in https://github.com/opencontainers/runc/pull/3850, just in case this invalid variable contains some sensitive information.

File: libcontainer/env.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kolyshkin) [reply libcontainer/env.go:46]: > len(key) +1? Or define val here using that and assign val?

Changed to `i+1` (might result in worse asm as it's not constant but who cares).

> I'm unsure about doiong this here vs in the system.Exec() func.

For both `runc start/run` and `runc init` we do explicit binary lookup (`exec.LookPath`) slightly earlier while we still have a chance to return an error (as we close everything before doing exec, we won't be able to properly return an error), and when we use a looked up binary path.

Now, if there was an `exec.LookPath` version which allows to specify PATH, we could use it instead of setting PATH in the current context. Alas, it doesn't exist and I don't want to replicate it in runc.

NOTE that currently (i.e. before this PR) we set *all* the environment variables from spec in the runc parent context (which can be bad in case someone would use GODEBUG or GOEXPERIMENT etc). This PR reduces *all* to PATH.

File: libcontainer/env.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+		}
+		if i == 0 {
+			return nil, false, errors.New("invalid environment variable: name cannot be empty")
+		}
+		key := kv[:i]
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply libcontainer/env.go:36]: Good question.

For the purpose of this PR, though,  I am not changing the behavior wrt environment validation, i.e. it stays as it was.

Looking into runtime-spec, it says "with the same semantics as [IEEE Std 1003.1-2008's `environ`](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_01). The latter kind of answers your question by saying "[o]ther characters may be permitted by an implementation; applications shall tolerate the presence of such names".

I was also looking into Go stdlib functions related to environment, and they do some minimal processing (pretty much the same as we do here, if I remember correctly). I _guess_ this is the way to go. If not, this should be raised via a separate issue.

File: libcontainer/env.go (hunk 5 above)
Comment (kolyshkin) [reply libcontainer/env.go:28]: > // Deduplication code based on dedupEnv from Go 1.22 os/exec.

As per the above comment, I stole this code from Go stdlib, and I don't want to change it from the original more than necessary.

File: libcontainer/env.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (rata) [reply libcontainer/env.go:31]: Ohh, makes sense!

File: libcontainer/env.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rata) [reply libcontainer/env.go]: Makes sense, thanks for the explanation!

File: libcontainer/env.go (hunk 5 above)
Comment (rata) [reply libcontainer/env.go:28]: Ohh, I didn't know that. Makes sense, thanks :)

File: libcontainer/env.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (rata) [reply libcontainer/env.go:36]: Thanks! If go is doing something very similar and that is what IEE says, I think it is fine to keep it like this.

**Review by rata — APPROVED** [review]
LGTM, thanks!

**Comment by kolyshkin** [comment]
OTOH I think this is ready for more reviews and potential inclusion.

**Comment by kolyshkin** [comment]
This is totally 1.3 material; let me rebase

**Comment by cyphar** [comment]
FWIW, I sent a patch to improve the performance of `os.Clearenv` (with the patch it's ~30% faster for an environment with 1000 variables set), though this is fixing other issues in addition to the `os.Clearenv` one. https://github.com/golang/go/pull/70672

**Comment by kolyshkin** [comment]
I've updated the patch (with somewhat better commit message and `prepareEnv` documentation. I think we can go ahead and merge this for v1.3, and leave the question raised in https://github.com/opencontainers/runc/pull/4325#issuecomment-2327591419 for later.

**Comment by lifubang** [comment]
> and leave the question raised in [#4325 (comment)](https://github.com/opencontainers/runc/pull/4325#issuecomment-2327591419) for later

But it's a break change for `StartContainer` hook, though this hook may be not wide spread used in practice.
Maybe we should add this break to `CHAGELOG` because we may forget it when making a release.

**Comment by cyphar** [comment]
> Maybe we should add this break to CHANGELOG because we may forget it when making a release.

I think we need to start adding things to the changelog in the implementation PR because we have missed breaking changes before (the most recent example I can think of is #3468, but I've definitely made the same mistake before as well -- it's very easy to skim over a patch when reading through the log to write the release notes).

File: libcontainer/standard_init_linux.go (hunk 4 above)
Comment (kolyshkin) [reply libcontainer/standard_init_linux.go]: Better late than never -- thank you @lifubang, I finally got it now.

This is now fixed and a test case is added.

I still prefer something like #4323 as it's much more straightforward.

**Comment by kolyshkin** [comment]
> > Maybe we should add this break to CHANGELOG because we may forget it when making a release.
> 
> I think we need to start adding things to the changelog in the implementation PR in general because we have missed breaking changes before (the most recent example I can think of is #3468, but I've definitely made the same mistake before as well -- it's very easy to skim over a patch when reading through the log to write the release notes).

@cyphar the above incompatibility issue reported by @lifubang is now fixed, but in the process I had to add a small change to libcontainer API, which is documented in the CHANGELOG now.

Doesn't mean this PR won't break something else, but at least I am not aware of anything.

File: libcontainer/standard_init_linux.go [hunk 9]
```diff
@@ -189,14 +190,22 @@ func (l *linuxStandardInit) Init() error {
...
 		return fmt.Errorf("can't restore pdeath signal: %w", err)
 	}
+
+	// In case we have any StartContainer hooks to run, and they don't
+	// have environment configured explicitly, make sure they will be run
+	// with the same environment as container's init.
+	if h := l.config.Config.Hooks[configs.StartContainer]; len(h) > 0 {
+		h.SetDefaultEnv(l.config.Env)
+	}
+
```
Comment (kolyshkin) [inline libcontainer/standard_init_linux.go:208]: @lifubang FYI this is the fix to the startContainer hook issue wrt environment you've reported back in June.

File: libcontainer/standard_init_linux.go [hunk 10]
```diff
@@ -189,14 +190,22 @@ func (l *linuxStandardInit) Init() error {
...
 	}
 	// finalizeNamespace can change user/group which clears the parent death
 	// signal, so we restore it here.
 	if err := pdeath.Restore(); err != nil {
 		return fmt.Errorf("can't restore pdeath signal: %w", err)
 	}
+
+	// In case we have any StartContainer hooks to run, and they don't
+	// have environment configured explicitly, make sure they will be run
+	// with the same environment as container's init.
```
Comment (kolyshkin) [inline libcontainer/standard_init_linux.go:204]: Maybe add a note that this behavior is not part of runtime-spec, but a de-facto historical thing we afraid to change.

File: libcontainer/standard_init_linux.go (hunk 9 above)
Comment (lifubang) [reply libcontainer/standard_init_linux.go:208]: Why added in here, but not together with the line before we prepare to run `StartContainer` hooks?
I mean put them before this line:
https://github.com/opencontainers/runc/blob/5a88394b1f68f1d5417b37485c9c26af91d1ccb6/libcontainer/standard_init_linux.go#L282

File: libcontainer/standard_init_linux.go (hunk 10 above)
Comment (lifubang) [reply libcontainer/standard_init_linux.go:204]: I think it worth to do it, because before this pr, it was doing in go-runtime, but now we have explicitly added the code to do this. We should let the reader know why we need this, and it may be removed in the future.

File: libcontainer/standard_init_linux.go (hunk 9 above)
Comment (kolyshkin) [reply libcontainer/standard_init_linux.go:208]: This is to minimize the amount of time spent in this last section of `Init` (in between writing `0` to exec fifo and the exec). As we've seen in #4437, there is a small race when runc thinks the container is running but in fact we're still doing some last preparations before exec.

File: libcontainer/standard_init_linux.go (hunk 10 above)
Comment (kolyshkin) [reply libcontainer/standard_init_linux.go:201]: @lifubang Added

File: CHANGELOG.md [hunk 11]
```diff
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### libcontainer API
+ * `configs.CommandHook` struct has changed, Command is now a pointer.
+   Also, `configs.NewCommandHook` now accepts a `*Command`. (#4325)
```
Comment (lifubang) [inline CHANGELOG.md:11]: Could you explain why we need to change this struct? Is it necessary?

Comment (lifubang) [reply CHANGELOG.md:11]: In fact, I think this PR should not be blocked by these small things, we should change as little as possible about the logic of hooks.

Comment (lifubang) [reply CHANGELOG.md:11]: Ah, it's about type assertion and the member var modification, make sense.

**Review by lifubang — APPROVED** [review]
LGTM
@opencontainers/runc-maintainers I think this one could be moved forward, would you have some other opinions?

File: CHANGELOG.md (hunk 11 above)
Comment (kolyshkin) [reply CHANGELOG.md:11]: Yes, otherwise it's not possible to modify the struct member (`Env`) in place, and the way the whole thing is structured, we only have final environment in a very late stage.

The alternatives are to either pass `env` to `RunHooks` as an additional argument, which I tried and it looks much worse.

File: libcontainer/init_linux.go [hunk 12]
```diff
@@ -541,11 +521,9 @@ func setupUser(config *initConfig) error {
 		return err
 	}
 
-	// if we didn't get HOME already, set it based on the user's HOME
-	if envHome := os.Getenv("HOME"); envHome == "" {
-		if err := os.Setenv("HOME", execUser.Home); err != nil {
```
Comment (lifubang) [inline libcontainer/init_linux.go:546]: Are there some situations which need this `HOME` env, just like `PATH` env?
We really set the `PATH` env if provided, please see:
https://github.com/opencontainers/runc/blob/97af673ac70a1279c75361dc2aeec44bda965cac/libcontainer/env.go#L49
So, maybe it's safe to set `HOME` env both in here and in `prepareEnv`? I don't know whether this scenario exists or not.

Comment (lifubang) [reply libcontainer/init_linux.go:546]: I have tested some scenarios to check that, it seems that this will not break anything.
So let's go ahead.

File: libcontainer/init_linux.go (hunk 3 above)
Comment (cyphar) [reply libcontainer/init_linux.go:190]: FWIW, now that the Go stdlib PR was merged (a year later...) this should be incredibly cheap.
